Make inline parsing customizable by providing an implemetaion of Inli…

lalunamel opened the pull request:
…neParser

Signed-off-by: Graham Bell gbell@pivotal.io
Signed-off-by: Cody Sehl csehl@pivotal.io

lalunamel commented:
Related to https://github.com/atlassian/commonmark-java/issues/68

robinst commented on commonmark/src/main/java/org/commonmark/parser/InlineParser.java:
```diff
@@ -14,4 +13,9 @@
+    /**
+     * @return how many characters were parsed as a reference, {@code 0} if none
+     */
+    int parseReference(String s);
```
I'm not sure if this belongs here or not, because it forces implementations of `InlineParser` to keep the reference map as state and it bakes this ugly part of the parser into the API.

For another issue, we'd also like to preserve parsed link reference definitions as nodes in the AST so that they can be used to render the original Markdown source back. So we'd have to break this API if we wanted to implement that.

Maybe I can look into that soon, but in the meantime, is there a way for you to not add this to the API while still enabling you to customize inline parsing?

lalunamel replied:
We've looked at the intended behavior behind parseReference and agree that it shouldn't be part of the InlineParser's API.
But it looks like the ParagraphParser uses this method when calling `closeBlock`.

We've searched through the code and we're not sure how to best divorce `parseReference` from the InlineParser interface when it's relied upon by ParagraphParser, given that the ParagraphParser is required to update the internal state/referenceMap of the InlineParser.

Any ideas here?

robinst replied:
We could put this method on a separate interface that lives in `.internal` for now, and `ParagraphParser` does an `if (inlineParser instanceof ReferenceParser)` before calling it. That way, we wouldn't have to put it in the API while your extension would still work I think. Thoughts?

That way this PR can go in while I think some more about what a proper `parseReference` should look like.

lalunamel commented:
Hey @robinst, any thoughts here?

lalunamel commented:
Ping!
(If we get this merged we don't need to manage a local, non-maven version of commonmark on all our dev machines :) )

robinst commented on commonmark/src/main/java/org/commonmark/parser/Parser.java:
```diff
@@ -75,12 +77,20 @@ public Node parse(String input) {
+        if (this.inlineParserFactory == null) {
+            return new InlineParserImpl(specialCharacters, delimiterCharacters, delimiterProcessors);
+        } else {
+            return this.inlineParserFactory.create();
```
Does your custom inline parser not need to know about `delimiterProcessors` etc? I think the `create` method should get a `InlineParserContext` (similar to `HtmlNodeRendererContext`), that looks like this:

```java
interface InlineParserContext {
    Map<Character, DelimiterProcessor> getDelimiterProcessors();
}
```

(The other two, `specialCharacters` and `delimiterCharacters` can be derived from the delimiter processors and are just an optimization, so don't need to be on the interface.)

lalunamel commented:
We split out `parseReference` into a separate `ReferenceParser` interface.
While not explicitly required, implementations of InlineParser require implementing both `InlineParser` and `ReferenceParser`, else the ParagraphParser won't work as expected.

We also changed the way custom `delimiterProcessor`s are specified for an InlineParser.
We didn't wrap them in a Context because we felt it was unnecessary.

robinst commented:
Thanks for making the changes, merged now. Will release with these changes soon!

robinst commented:
I just released 0.9.0 which includes these changes: https://github.com/atlassian/commonmark-java/releases/tag/commonmark-parent-0.9.0

(Was waiting for another PR to include in the release, but it's not ready yet.)
